power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -583,6 +583,27 @@ template <symmetry_tag sym> struct NodeState {
     }
 };
 
+class RankIterator {
+  public:
+    RankIterator() = default;
+
+    void iterate_ranks(auto const& regulator_order, auto adjust_transformer_in_rank, bool& tap_changed,
```

Comment:
since this logic is actually more generic than regulated transformers, maybe the names here can also be more generic, e.g.:
* `regulator_order` -> `ranked_order`
* `adjust_transformer_in_rank` -> `apply_single_object_in_rank`
* `tap_changed` -> `adjusted` or `additive_result`